apps/capi/test/capi_base_api_token_tests_SUITE.erl

```diff
@@ -1713,3 +1732,7 @@ check_support_decrypt_v2_test(_Config) ->
         PaymentTool
     ),
     ?assertEqual(<<"2020-10-29T23:44:15.499Z">>, capi_utils:deadline_to_binary(ValidUntil)).
+
+-spec throwing_reply(any()) -> {ok, term()}.
+throwing_reply(Error) ->
+    {ok, erlang:apply(erlang, throw, [Error])}.
```

Comment:
Это выглядит как хак.

Reply:
> Это выглядит как хак.

да - это хак против глубинного диалайзера. в противном случае в 11 местах нужно ставить атрибуты вида  `-dialyzer({nowarn_function, function_name/1}).` при этом рискуя подавить нужные предупреждения.  я долго провозился, но более прозрачного  решения не нашел.

Reply:
Выглядит так, словно это какая-то проблема в диалайзере. Он игнорирует даже явное указание что `capi_ct_helper:mock_services/2` принимает функцию без выхода. Вероятно это связано с тем, что тут используются безымянные функции.

Можно тогда сделать, например так
```erlang
diff --git a/apps/capi/test/capi_base_api_token_tests_SUITE.erl b/apps/capi/test/capi_base_api_token_tests_SUITE.erl
index cceff64..1af2dba 100644
--- a/apps/capi/test/capi_base_api_token_tests_SUITE.erl
+++ b/apps/capi/test/capi_base_api_token_tests_SUITE.erl
@@ -311,7 +311,7 @@ create_invoice_autorization_error_test(Config) ->
     _ = capi_ct_helper:mock_services(
         [
             {invoicing, fun('Create', {_, #payproc_InvoiceParams{party_id = <<"WrongPartyID">>}}) ->
-                throwing_reply(#payproc_InvalidUser{})
+                {error, #payproc_InvalidUser{}}
             end},
             {generator, fun('GenerateID', _) -> capi_ct_helper_bender:generate_id(<<"bender_key">>) end}
         ],
@@ -384,7 +384,7 @@ create_invoice_template_autorization_error_test(Config) ->
                 'Create',
                 {_, #payproc_InvoiceTemplateCreateParams{party_id = <<"WrongPartyID">>}}
             ) ->
-                throwing_reply(#payproc_InvalidUser{})
+                {error, #payproc_InvalidUser{}}
             end}
         ],
         Config
@@ -484,7 +484,7 @@ create_customer_autorization_error_test(Config) ->
     _ = capi_ct_helper:mock_services(
         [
             {customer_management, fun('Create', {#payproc_CustomerParams{party_id = <<"WrongPartyID">>}}) ->
-                throwing_reply(#payproc_InvalidUser{})
+                {error, #payproc_InvalidUser{}}
             end}
         ],
         Config
@@ -596,11 +596,11 @@ create_refund_error(Config) ->
         [
             {invoicing, fun
                 ('RefundPayment', {_, <<"42">>, _, _}) ->
-                    throwing_reply(#payproc_InvalidPartyStatus{
+                    {error, #payproc_InvalidPartyStatus{
                         status = {blocking, {blocked, #domain_Blocked{reason = ?STRING, since = ?TIMESTAMP}}}
-                    });
+                    }};
                 ('RefundPayment', {_, <<"43">>, _, _}) ->
-                    throwing_reply(#payproc_InvalidContractStatus{status = {expired, #domain_ContractExpired{}}})
+                    {error, #payproc_InvalidContractStatus{status = {expired, #domain_ContractExpired{}}}}
             end},
             {generator, fun('GenerateID', _) -> capi_ct_helper_bender:generate_id(BenderKey) end}
         ],
@@ -798,7 +798,7 @@ get_shop_by_id_for_party_error_test(Config) ->
     _ = capi_ct_helper:mock_services(
         [
             {party_management, fun('GetShop', {_, <<"WrongPartyID">>, _}) ->
-                throwing_reply(#payproc_InvalidUser{})
+                {error, #payproc_InvalidUser{}}
             end}
         ],
         Config
@@ -826,7 +826,7 @@ get_shops_for_party_ok_test(Config) ->
 -spec get_shops_for_party_error_test(config()) -> _.
 get_shops_for_party_error_test(Config) ->
     _ = capi_ct_helper:mock_services(
-        [{party_management, fun('Get', {_, <<"WrongPartyID">>}) -> throwing_reply(#payproc_InvalidUser{}) end}],
+        [{party_management, fun('Get', {_, <<"WrongPartyID">>}) -> {error, #payproc_InvalidUser{}} end}],
         Config
     ),
     ?assertMatch(
@@ -854,7 +854,7 @@ activate_shop_for_party_error_test(Config) ->
     _ = capi_ct_helper:mock_services(
         [
             {party_management, fun('ActivateShop', {_, <<"WrongPartyID">>, _}) ->
-                throwing_reply(#payproc_InvalidUser{})
+                {error, #payproc_InvalidUser{}}
             end}
         ],
         Config
@@ -884,7 +884,7 @@ suspend_shop_for_party_error_test(Config) ->
     _ = capi_ct_helper:mock_services(
         [
             {party_management, fun('SuspendShop', {_, <<"WrongPartyID">>, _}) ->
-                throwing_reply(#payproc_InvalidUser{})
+                {error, #payproc_InvalidUser{}}
             end}
         ],
         Config
@@ -1209,7 +1209,7 @@ create_webhook_limit_exceeded_test(Config) ->
     _ = capi_ct_helper:mock_services(
         [
             {party_management, fun('GetShop', _) -> {ok, ?SHOP} end},
-            {webhook_manager, fun('Create', _) -> throwing_reply(#webhooker_LimitExceeded{}) end}
+            {webhook_manager, fun('Create', _) -> {error, #webhooker_LimitExceeded{}} end}
         ],
         Config
     ),
@@ -1575,7 +1575,7 @@ check_no_payment_by_external_id_test(Config) ->
     BenderContext = capi_msgp_marshalling:marshal(#{<<"context_data">> => #{<<"invoice_id">> => <<"123">>}}),
     _ = capi_ct_helper:mock_services(
         [
-            {invoicing, fun('GetPayment', _) -> throwing_reply(#payproc_InvoicePaymentNotFound{}) end},
+            {invoicing, fun('GetPayment', _) -> {error, #payproc_InvoicePaymentNotFound{}} end},
             {bender, fun('GetInternalID', _) ->
                 InternalKey = capi_utils:get_unique_id(),
                 {ok, capi_ct_helper_bender:get_internal_id_result(InternalKey, BenderContext)}
@@ -1615,7 +1615,7 @@ check_no_internal_id_for_external_id_test(Config) ->
     ExternalID = capi_utils:get_unique_id(),
     _ = capi_ct_helper:mock_services(
         [
-            {bender, fun('GetInternalID', _) -> throwing_reply(capi_ct_helper_bender:no_internal_id()) end}
+            {bender, fun('GetInternalID', _) -> {error, capi_ct_helper_bender:no_internal_id()} end}
         ],
         Config
     ),
@@ -1732,7 +1732,3 @@ check_support_decrypt_v2_test(_Config) ->
         PaymentTool
     ),
     ?assertEqual(<<"2020-10-29T23:44:15.499Z">>, capi_utils:deadline_to_binary(ValidUntil)).
-
--spec throwing_reply(any()) -> {ok, term()}.
-throwing_reply(Error) ->
-    {ok, erlang:apply(erlang, throw, [Error])}.
diff --git a/apps/capi/test/capi_ct_helper.erl b/apps/capi/test/capi_ct_helper.erl
index 0a391eb..c6ba50f 100644
--- a/apps/capi/test/capi_ct_helper.erl
+++ b/apps/capi/test/capi_ct_helper.erl
@@ -30,6 +30,11 @@
 -define(CAPI_URL, ?CAPI_HOST_NAME ++ ":" ++ integer_to_list(?CAPI_PORT)).
 
 %%
+
+-type services() :: [service()].
+-type service() :: {Name :: atom(), Handler :: service_handler()}.
+-type service_handler() :: fun((woody:func(), woody:args()) -> {ok, woody:result()} | {error, term()}).
+
 -type config() :: [{atom(), any()}].
 -type app_name() :: atom().
 
@@ -174,7 +179,7 @@ start_mocked_service_sup(Module) ->
 stop_mocked_service_sup(SupPid) ->
     exit(SupPid, shutdown).
 
--spec mock_services(_, _) -> _.
+-spec mock_services(services(), _) -> _.
 mock_services(Services, SupOrConfig) ->
     {BenderClientServices, WoodyServices} = lists:partition(
         fun({ServiceName, _}) ->
diff --git a/apps/capi/test/capi_dummy_service.erl b/apps/capi/test/capi_dummy_service.erl
index 767aa4f..af9c855 100644
--- a/apps/capi/test/capi_dummy_service.erl
+++ b/apps/capi/test/capi_dummy_service.erl
@@ -7,4 +7,9 @@
 -spec handle_function(woody:func(), woody:args(), woody_context:ctx(), woody:options()) ->
     {ok, woody:result()} | no_return().
 handle_function(FunName, Args, _, #{function := Fun}) ->
-    Fun(FunName, Args).
+    case Fun(FunName, Args) of
+        {ok, _Reply} = Result ->
+            Result;
+        {error, Exception} ->
+            erlang:throw(Exception)
+    end.
```

Ну или отключить првоерку тестов. Оставлять такие костыли мне кажется неправильным.

Reply:
> Ну или отключить првоерку тестов. Оставлять такие костыли мне кажется неправильным.

не очень согласен. поиск переопределения нестандартного способа обработки ответа где-то в недрах подменного хендлера это очень сложно для работы с кодом. вынос исключения в функцию throwing_reply это очень понятный и прозрачный способ. 
тут осталась одна проблема, что диалайзер игнорирует директиву подавления -no-return. т.е. это очевидный баг для анонимных функций. потому спрятать за временным решением apply -  норм решением, учитывая, что костыль в зоне прямой видимости. 
ps. кстати, я еще не проверял на 23.2.3 - там многие хаки перестали быть нужны. проверю и отпишусь.

Reply:
@ciiol >Ну или отключить првоерку тестов. Оставлять такие костыли мне кажется неправильным.

обновление не помого, потому-что проект бы уже обновлен. поэтому применил решение с переносом исключения в capi_dummy_service